site/public/js/autograding-status.js

```diff
@@ -0,0 +1,132 @@
+                table = document.getElementById('machine-table');
+                $('#machine-table tbody').html('');
+                Object.keys(json.ongoing_job_info).forEach(key => {
+                    const info = json.ongoing_job_info[key];
+                    info.forEach(elem => {
+                        const new_row = table.getElementsByTagName('tbody')[0].insertRow(-1);
+                        new_row.insertCell().innerHTML = key;
+                        new_row.insertCell().innerHTML = elem.semester;
+                        new_row.insertCell().innerHTML = elem.course;
+                        new_row.insertCell().innerHTML = elem.gradeable_id;
+                        new_row.insertCell().innerHTML = elem.user_id;
+                        new_row.insertCell().innerHTML = elem.elapsed_time;
+                        new_row.insertCell().innerHTML = elem.error;
+                    });
+                });
+
+                // Update Grading Monitor table
+                table = document.getElementById('autograding-status-table');
+                const new_row = table.insertRow(3);
+                let new_cell = new_row.insertCell();
+                new_cell.innerHTML = json.time;
+                new_cell.className = 'right-boarder';
+                new_row.insertCell().innerHTML = json.queue_counts.interactive_ongoing;
+                new_cell = new_row.insertCell();
+                new_cell.innerHTML = json.queue_counts.interactive;
+                new_cell.className = 'right-boarder';
+                new_row.insertCell().innerHTML = json.queue_counts.regrade_ongoing;
+                new_cell = new_row.insertCell();
+                new_cell.innerHTML = json.queue_counts.regrade;
+                new_cell.className = 'right-boarder';
+                Object.keys(json.machine_grading_counts).forEach((key, i) => {
+                    if (i == Object.keys(json.machine_grading_counts).length - 1) {
+                        new_cell = new_row.insertCell();
+                        new_cell.innerHTML = json.machine_grading_counts[key];
+                        new_cell.className = 'right-boarder';
+                    }
+                    else {
+                        new_row.insertCell().innerHTML = json.machine_grading_counts[key];
+                    }
+                });
+                Object.keys(json.capability_queue_counts).forEach(key => {
+                    new_row.insertCell().innerHTML = json.capability_queue_counts[key];
+                });
+                // Check if old logs should be removed to make room for new logs
+                if ($('#autograding-status-table tbody tr').length > max_log) {
+                    // +3 to account for the thead
+                    table.deleteRow(max_log + 3);
+                }
+
+                // Queue this function to be run again after specified delay
+                interval = setInterval(updateTable, refresh_freq);
```

Comment:
Instead of using setInterval, it might be better to use setTimeout() and then call that recursively, waiting for requests to come back before sending out the next one.  That way, if requests are taking a long time, we don't spam the server with requests.

Reply:
I considered using `setTimeout()`, but then the button for pausing the timeout becomes tricky, or at least I don't know of a way to get rid of things queued to run with `setTimeout()`. With setInterval, however, I can use the `clearInterval()`.

Reply:
An easy way to do this would be to use the button to add/remove an attribute somewhere and then check for the existence of that attribute before calling `setTimeout()` again.

Reply:
What are the advantages of using `setTimeout()` over `setInterval()`, right now `setInterval()` also allows us to wait for the response to come back before sending another request.

Reply:
It would just mean that you wouldn't have to always be clearing/resetting the interval.  If I understand the code correctly, this code essentially uses `setInterval()` to do the job of `setTimeout()`.  `setTimeout()` only runs once after a set amount of time, while the current code sets the interval and runs the function once which clears the interval and then re-sets it.